lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2382 @@
+            "observation.state": deque(maxlen=self.config.n_obs_steps),
+            "action": deque(maxlen=self.config.n_action_pred_chunk),
+        }
+
+    @torch.no_grad
+    def select_action(self, batch: dict[str, Tensor]) -> Tensor:
+        """Select a single action given environment observations.
+
+        This method wraps `select_actions` in order to return one action at a time for execution in the
+        environment. It works by managing the actions in a queue and only calling `select_actions` when the
+        queue is empty.
+        """
+
+        self.eval()
+
+        batch = self.normalize_inputs(batch)
+        self._queues = populate_queues(self._queues, batch)
+
+        if not self.check_discretized():
+            self.vqbet.action_head.vqvae_model.discretized = True
+            # VQ-BeT can predict action only after finishing action discretization.
+            # We added a logit to force self.vqbet.action_head.vqvae_model.discretized to be True if not self.check_discretized() to account for the case of predicting with a pretrained model, but this shouldn't happen if you're learning from scratch, so set eval_freq greater than discretize_step.
+            warnings.warn('To evaluate in the environment, the model was forced to stop learning the Residual VQ. If you are not evaluating with a pre-trained model, this can degrade overall performance. To avoid this warning, please set "eval_freq" greater than "discretize_step".')
+        assert "observation.image" in batch
+        assert "observation.state" in batch
+
+        if len(self._queues["action"]) == 0:
+
+            batch = {key: torch.stack(list(self._queues[key]), dim=1) for key in batch}
+            actions = self.vqbet(batch, rollout=True)[:, : self.config.n_action_pred_chunk]
+
+            # the dimension of returned action is (batch_size, n_action_pred_chunk, action_dim)
+            actions = self.unnormalize_outputs({"action": actions})["action"]
+            # since the data in the action queue's dimension is (n_action_pred_chunk, batch_size, action_dim, we transpose the action and fill the queue
+            self._queues["action"].extend(actions.transpose(0, 1))
+
+        action = self._queues["action"].popleft()
+        return action
+
+    def forward(self, batch: dict[str, Tensor]) -> dict[str, Tensor]:
+        """Run the batch through the model and compute the loss for training or validation."""
+        batch = self.normalize_inputs(batch)
+        batch = self.normalize_targets(batch)
+        # VQ-BeT discretizes action using VQ-VAE before training BeT (please refer to section 3.2 in the VQ-BeT paper https://arxiv.org/pdf/2403.03181)
+        if not self.check_discretized():
+            loss, n_different_codes, n_different_combinations = self.vqbet.discretize(self.config.discretize_step, batch['action'])
+            return {"loss": loss, "n_different_codes": n_different_codes, "n_different_combinations": n_different_combinations}
```

Comment:
added explanations:)

Reply:
Thanks! I _think_ I understand this. Can you let me know if my understanding is correct?

```
n_different_codes: how many of the total possible VQ codes are being used (how many of them have at least one encoder embedding as a nearest neighbor). This can be at most `vqvae_n_embed`.
n_different_combinations: how many different code combinations are being used out of all possible combinations. This can be at most `vqvae_n_embed ^ vqvae_groups` (hint consider the RVQ as a decision tree).

But shouldn't `n_different_codes` max out at `vqvae_n_embed * vqvae_groups`? That's how many codes there are in total. Or are you only referring to the codes of the first RVQ layer?

Btw: I think this is a great metric to track!

Reply:
Exactly. The reason why the max value of n_different_codes is `vqvae_n_embed` is we don't differentiated the code of different RVQ layers. But after your comment, I agree that we should differentiate them to calculate `n_different_codes`. Thus, I changed the `n_different_codes` calculation part. 

from
```
n_different_codes = len(torch.unique(metric[2]))
```
to
```
n_different_codes = sum([len(torch.unique(metric[2][:, i])) for i in range(vqvae_model.vqvae_num_layers)])
```

Also, thanks to your more detailed explanations, I updated the comments.

```
# n_different_codes: how many of the total possible VQ codes are being used in single batch (how many of them have at least one encoder embedding as a nearest neighbor). This can be at most `vqvae_n_embed * number of layers of RVQ (=2)`.
# n_different_combinations: how many different code combinations are being used out of all possible combinations in single batch. This can be at most `vqvae_n_embed ^ number of layers of RVQ (=2)` (hint consider the RVQ as a decision tree).
```